k8s: Use kubelet's logic to close all idle connections

christarazi opened the pull request:
Following up from https://github.com/cilium/cilium/pull/19259, this
commit implements the same logic as kubelet [1] to close all idle
connections. We can rely on HTTP/2 to determine which connections are
idle (HTTP/2 pings), instead of closing all connections indiscriminately
(when the heartbeat fails) via the custom dialer in setDialer().

We can assume HTTP/2 support as it has been supported since K8s 1.20
[2]. Users can disable HTTP/2 by setting the DISABLE_HTTP2 environment
variable, however this only really meant from very advanced users, hence
no documentation. Kubernetes has been running with HTTP/2 as the default
for a long time (~since Dec 2020) without major issues.

CloseIdleConnectionsFor() was introduced in K8s 1.23 [3], hence this commit
cannot be backported to any earlier version of Cilium which doesn't
support K8s 1.23.

[1]: https://github.com/kubernetes/kubernetes/pull/104844
[2]: https://github.com/kubernetes/kubernetes/pull/95981/
[3]: https://github.com/kubernetes/kubernetes/commit/b9d865a8185b62d83e9ff81b0e3499a26ac6960d

Suggested-by: André Martins <andre@cilium.io>
Signed-off-by: Chris Tarazi <chris@isovalent.com>

aanm commented on pkg/k8s/init.go:
```diff
@@ -139,7 +141,7 @@ func Init(conf k8sconfig.Configuration) error {
 	if err != nil {
 		return fmt.Errorf("unable to create k8s client rest configuration: %s", err)
 	}
-	closeAllDefaultClientConns := setDialer(restConfig)
```
This line should not be removed, the `restConfig` is being modified and the result of it is being used on the following lines. If we don't modify the `restConfig` here then the none of the clients will use the dialer set by `setDialer`

christarazi replied:
Fixed, good catch.

aanm replied:
It's not fixed, `getCloseAllConnsFunc(restConfig)` is only executed on line 186. `setDialer` needs to be executed on this line.

christarazi replied:
Should be good now

aanm commented on pkg/k8s/init.go:
```diff
@@ -164,6 +166,18 @@ func Init(conf k8sconfig.Configuration) error {
+			utilnet.CloseIdleConnectionsFor(restConfig.Transport)
+		}
+	}
+
```
This needs to be done before line 150

christarazi committed:
k8s: Use kubelet's logic to close all idle connections

Following up from https://github.com/cilium/cilium/pull/19259, this
commit implements the same logic as kubelet [1] to close all idle
connections. We can rely on HTTP/2 to determine which connections are
idle (HTTP/2 pings), instead of closing all connections indiscriminately
(when the heartbeat fails) via the custom dialer in setDialer().

We can assume HTTP/2 support as it has been supported since K8s 1.20
[2]. Users can disable HTTP/2 by setting the DISABLE_HTTP2 environment
variable, however this only really meant from very advanced users, hence
no documentation. Kubernetes has been running with HTTP/2 as the default
for a long time (~since Dec 2020) without major issues.

CloseIdleConnectionsFor() was introduced in K8s 1.23 [3], hence this commit
cannot be backported to any earlier version of Cilium which doesn't
support K8s 1.23.

[1]: https://github.com/kubernetes/kubernetes/pull/104844
[2]: https://github.com/kubernetes/kubernetes/pull/95981/
[3]: https://github.com/kubernetes/kubernetes/commit/b9d865a8185b62d83e9ff81b0e3499a26ac6960d

Suggested-by: André Martins <andre@cilium.io>
Signed-off-by: Chris Tarazi <chris@isovalent.com>
